### PR TITLE
alpha-UCB bandit operator selection scheme

### LINKED ISSUE #73: Bandit weight scheme
Our current weight schemes are all very simple. I feel more can be learned, perhaps using some sort of multi-armed bandit.

### PR DESCRIPTION
**Pull request by N-Wouda**
Closes #73. This PR:
- [x] Fixes the argument order of other operator schemes, and their tests and use in examples
- [x] Implements $\alpha$-UCB, a bandit algorithm for operator selection
- [x] Adds tests for $\alpha$-UCB
- [ ] Uses $\alpha$-UCB in some example [EDIT: postponed for later; with a new example]

### CODE REVIEW

**Comment by N-Wouda** [comment]
@leonlan any ideas for which example to use this in?

**Comment by N-Wouda** [comment]
>We could augment the CVRP example to include more destroy/repair operators. This is fairly easy. Or I could create a new example for the permutation flow shop problem that uses 2x2 operators: block removal, random removal, greedy insert, random insert. A little bit more work.

TSP has 3x1, CSP 2x2 (but that's not a really interesting problem), RCPSP 3x1. We could add another repair operator to either RCPSP or TSP, and then call it a day? Maybe greedy insert or some such for RCPSP (the TSP example is already pretty good, but the RCPSP can probably be improved).

A whole new example is even nicer, but I can imagine that's a bit more work that we do not want to put in right now.

File: alns/ALNS.py [hunk 1]
```diff
@@ -81,7 +81,9 @@ def repair_operators(self) -> List[Tuple[str, _OperatorType]]:
         """
         return list(self._r_ops.items())
 
-    def add_destroy_operator(self, op: _OperatorType, name: str = None):
+    def add_destroy_operator(
+        self, op: _OperatorType, name: Optional[str] = None
+    ):
```
Comment (N-Wouda) [inline alns/ALNS.py:86]: Unrelated to this PR, but mypy has finally started complaining about this (that's why master is failing)

File: .github/workflows/alns.yml [hunk 2]
```diff
@@ -17,9 +17,9 @@ jobs:
       matrix:
```
Comment (N-Wouda) [inline .github/workflows/alns.yml:17]: The changes in this file fix a few warnings about deprecations in the CI runner.

**Comment by N-Wouda** [comment]
I've added $\alpha$-UCB to the features notebook, so we at least cover it in our tour of the package features. But a more realistic example would still be nice to have.

File: alns/select/AlphaUCB.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        return self._scores
+
+    @property
+    def alpha(self) -> float:
+        return self._alpha
+
+    def __call__(self, rnd, best, curr):
+        """
+        Returns the (destroy, repair) operator pair that maximises the average
+        reward and exploration bonus.
```
Comment (leonlan) [inline alns/select/AlphaUCB.py:102]: Suggestion: perhaps add a small note in the class docstring about the Q value being a sum of the avg reward and exploration bonus? The Q expression looked quite complex when I read it first, but knowing that its just avg reward + exploration bonus immediately makes sense.

File: alns/ALNS.py (hunk 1 above)
Comment (N-Wouda) [reply alns/ALNS.py:86]: We don't fix dependency versions, so I think we just got a more recent mypy that complains.

At some point we should probably explicitly test the lowest supported dependency versions in a dedicated CI run, but I haven't gotten around to that yet.

**Review by leonlan — APPROVED** [review]
LGTM.

**Pull request merged**